File: android/src/main/java/com/easemob/im_flutter_sdk/EMPushManagerWrapper.java
package com.easemob.im_flutter_sdk;

import android.content.Context;

import com.hyphenate.chat.EMClient;
import com.hyphenate.chat.EMConversation;
import com.hyphenate.chat.EMGroup;
import com.hyphenate.chat.EMPushConfigs;
import com.hyphenate.chat.EMPushManager.DisplayStyle;
import com.hyphenate.chat.EMSilentModeParam;
import com.hyphenate.chat.EMSilentModeResult;
import com.hyphenate.exceptions.HyphenateException;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;


public class EMPushManagerWrapper extends EMWrapper implements MethodCallHandler {

    EMPushManagerWrapper(FlutterPlugin.FlutterPluginBinding flutterPluginBinding, String channelName) {
        super(flutterPluginBinding, channelName);
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        JSONObject param = (JSONObject)call.arguments;
        try {
            if (EMSDKMethod.getImPushConfig.equals(call.method)) {
                getImPushConfig(param, call.method, result);
            }
            else if(EMSDKMethod.getImPushConfigFromServer.equals(call.method)){
                getImPushConfigFromServer(param, call.method, result);
            }
            else if(EMSDKMethod.updatePushNickname.equals(call.method)){
                updatePushNickname(param, call.method, result);
            }
            else if(EMSDKMethod.updateImPushStyle.equals(call.method)){
                updateImPushStyle(param, call.method, result);
            }
            else if(EMSDKMethod.updateGroupPushService.equals(call.method)){
                updateGroupPushService(param, call.method, result);
            }
            else if(EMSDKMethod.updateHMSPushToken.equals(call.method)){
                updateHMSPushToken(param, call.method, result);
            }
            else if(EMSDKMethod.updateFCMPushToken.equals(call.method)){
                updateFCMPushToken(param, call.method, result);
            }
            else if (EMSDKMethod.enableOfflinePush.equals(call.method)) {
                enableOfflinePush(param, call.method, result);
            }
            else if (EMSDKMethod.disableOfflinePush.equals(call.method)){
                disableOfflinePush(param, call.method, result);
            }
            else if (EMSDKMethod.getNoPushGroups.equals(call.method)) {
                getNoPushGroups(param, call.method, result);
            }
            else if (EMSDKMethod.updateUserPushService.equals(call.method)) {
                updateUserPushService(param, call.method, result);
            }
            else if (EMSDKMethod.reportPushAction.equals(call.method)) {
                reportPushAction(param, call.method, result);
            }
            else if (EMSDKMethod.setConversationSilentMode.equals(call.method)) {
                setConversationSilentMode(param, call.method, result);
            }
            else if (EMSDKMethod.removeConversationSilentMode.equals(call.method)) {
                removeConversationSilentMode(param, call.method, result);
            }
            else if (EMSDKMethod.fetchConversationSilentMode.equals(call.method)) {
                fetchConversationSilentMode(param, call.method, result);
            }
            else if (EMSDKMethod.setSilentModeForAll.equals(call.method)) {
                setSilentModeForAll(param, call.method, result);
            }
            else if (EMSDKMethod.fetchSilentModeForAll.equals(call.method)) {
                fetchSilentModeForAll(param, call.method, result);
            }
            else if (EMSDKMethod.fetchSilentModeForConversations.equals(call.method)) {
                fetchSilentModeForConversations(param, call.method, result);
            }
            else if (EMSDKMethod.setPreferredNotificationLanguage.equals(call.method)) {
                setPreferredNotificationLanguage(param, call.method, result);
            }
            else if (EMSDKMethod.fetchPreferredNotificationLanguage.equals(call.method)) {
                fetchPreferredNotificationLanguage(param, call.method, result);
            }
            else if (EMSDKMethod.getPushTemplate.equals(call.method)) {
                getPushTemplate(param, call.method, result);
            }
            else if (EMSDKMethod.setPushTemplate.equals(call.method)) {
                setPushTemplate(param, call.method, result);
            }
            else {
                super.onMethodCall(call, result);
            }
        }catch (JSONException e) {

        }
    }

    private void getImPushConfig(JSONObject params, String channelName,  Result result) throws JSONException {
        asyncRunnable(()->{
            EMPushConfigs configs = EMClient.getInstance().pushManager().getPushConfigs();
            onSuccess(result, channelName, EMPushConfigsHelper.toJson(configs));
        });

    }

    private void getImPushConfigFromServer(JSONObject params, String channelName,  Result result) throws JSONException {
        asyncRunnable(()->{
            try {
                EMPushConfigs configs = EMClient.getInstance().pushManager().getPushConfigsFromServer();
                onSuccess(result, channelName, EMPushConfigsHelper.toJson(configs));
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void updatePushNickname(JSONObject params, String channelName,  Result result) throws JSONException {
        String nickname = params.getString("nickname");

        asyncRunnable(()->{
            try {
                EMClient.getInstance().pushManager().updatePushNickname(nickname);
                onSuccess(result, channelName, nickname);
            } catch (HyphenateException e) {
                onError(result, e);
            }
        });
    }


    private void enableOfflinePush(JSONObject params, String channelName, Result result) throws JSONException
    {
        asyncRunnable(()-> {
            try {
                EMClient.getInstance().pushManager().enableOfflinePush();
                onSuccess(result, channelName, null);
            } catch(HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void disableOfflinePush(JSONObject params, String channelName, Result result) throws JSONException
    {
        int startTime = params.getInt("start");
        int endTime = params.getInt("end");
        asyncRunnable(()-> {
            try {
                EMClient.getInstance().pushManager().disableOfflinePush(startTime, endTime);
                onSuccess(result, channelName, null);
            } catch(HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void getNoPushGroups(JSONObject params, String channelName, Result result)  throws JSONException {
        asyncRunnable(()-> {
            List<String> groups = EMClient.getInstance().pushManager().getNoPushGroups();
            onSuccess(result, channelName, groups);
        });
    }

    private void getNoPushUsers(JSONObject params, String channelName, Result result) throws JSONException {
        asyncRunnable(()->{
            List<String> list = EMClient.getInstance().pushManager().getNoPushUsers();
            onSuccess(result, channelName, list);
        });
    }

    private void updateImPushStyle(JSONObject params, String channelName,  Result result) throws JSONException {
        DisplayStyle style = params.getInt("pushStyle") == 0 ? DisplayStyle.SimpleBanner : DisplayStyle.MessageSummary;
        EMClient.getInstance().pushManager().asyncUpdatePushDisplayStyle(style, new EMWrapperCallBack(result, channelName, true));
    }

    private void updateGroupPushService(JSONObject params, String channelName,  Result result) throws JSONException {
        JSONArray groupIds = params.getJSONArray("group_ids");
        boolean noPush = params.getBoolean("noPush");

        List<String> groupList = new ArrayList<>();
        for (int i = 0; i < groupIds.length(); i++) {
            String groupId = groupIds.getString(i);
            groupList.add(groupId);
        }
        asyncRunnable(()-> {
            try {
                EMClient.getInstance().pushManager().updatePushServiceForGroup(groupList, noPush);
                onSuccess(result, channelName, null);
            } catch(HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void updateUserPushService(JSONObject params, String channelName, Result result) throws JSONException {
        JSONArray groupIds = params.getJSONArray("user_ids");
        boolean noPush = params.getBoolean("noPush");

        List<String> userList = new ArrayList<>();
        for (int i = 0; i < groupIds.length(); i++) {
            String userId = groupIds.getString(i);
            userList.add(userId);
        }
        asyncRunnable(()-> {
            try {
                EMClient.getInstance().pushManager().updatePushServiceForUsers(userList, noPush);
                onSuccess(result, channelName, null);
            } catch(HyphenateException e) {
                onError(result, e);
            }
        });
    }

    private void updateHMSPushToken(JSONObject params, String channelName,  Result result) throws JSONException {
        String token = params.getString("token");
        asyncRunnable(()->{
            EMClient.getInstance().sendHMSPushTokenToServer(token);
            onSuccess(result, channelName, token);
        });
    }

    private void updateFCMPushToken(JSONObject params, String channelName,  Result result) throws JSONException {
        String token = params.getString("token");
        String fcmKey = EMClient.getInstance().getOptions().getPushConfig().getFcmSenderId();
        EMClient.getInstance().pushManager().bindDeviceToken(fcmKey, token, new EMWrapperCallBack(result, channelName, null));
    }

    private void reportPushAction(JSONObject params, String channelName, Result result) throws JSONException {

    }

    private void setConversationSilentMode(JSONObject params, String channelName, Result result) throws JSONException {
        String conversationId = params.getString("conversationId");
        EMConversation.EMConversationType type = EMConversationHelper.typeFromInt(params.getInt("conversationType"));
        EMSilentModeParam param = EMSilentModeParamHelper.fromJson(params.getJSONObject("param"));
        EMClient.getInstance().pushManager().setSilentModeForConversation(conversationId, type, param, new EMValueWrapperCallBack<EMSilentModeResult>(result, channelName){
            @Override
            public void onSuccess(EMSilentModeResult object) {
                super.updateObject(null);
            }
        });
    }
    private void removeConversationSilentMode(JSONObject params, String channelName, Result result) throws JSONException {
        String conversationId = params.getString("conversationId");
        EMConversation.EMConversationType type = EMConversationHelper.typeFromInt(params.getInt("conversationType"));
        EMClient.getInstance().pushManager().clearRemindTypeForConversation(conversationId, type, new EMWrapperCallBack(result, channelName, null));
    }

    private void fetchConversationSilentMode(JSONObject params, String channelName, Result result) throws JSONException {
        String conversationId = params.getString("conversationId");
        EMConversation.EMConversationType type = EMConversationHelper.typeFromInt(params.getInt("conversationType"));
        EMClient.getInstance().pushManager().getSilentModeForConversation(conversationId, type, new EMValueWrapperCallBack<EMSilentModeResult>(result, channelName){
            @Override
            public void onSuccess(EMSilentModeResult object) {
                super.updateObject(EMSilentModeResultHelper.toJson(object));
            }
        });
    }

    private void setSilentModeForAll(JSONObject params, String channelName, Result result) throws JSONException {
        EMSilentModeParam param =  EMSilentModeParamHelper.fromJson(params.getJSONObject("param"));
        EMClient.getInstance().pushManager().setSilentModeForAll(param ,new EMValueWrapperCallBack<EMSilentModeResult>(result, channelName){
            @Override
            public void onSuccess(EMSilentModeResult object) {
                super.updateObject(null);
            }
        });
    }

    private void fetchSilentModeForAll(JSONObject params, String channelName, Result result) throws JSONException {
        EMClient.getInstance().pushManager().getSilentModeForAll(new EMValueWrapperCallBack<EMSilentModeResult>(result, channelName){
            @Override
            public void onSuccess(EMSilentModeResult object) {
                super.updateObject(EMSilentModeResultHelper.toJson(object));
            }
        });
    }
    private void fetchSilentModeForConversations(JSONObject params, String channelName, Result result) throws JSONException {
        Iterator iterator = params.keys();
        ArrayList<EMConversation> list = new ArrayList<>();
        while (iterator.hasNext()) {
            String conversationId = (String)iterator.next();
            EMConversation.EMConversationType type = EMConversationHelper.typeFromInt(params.getInt(conversationId));
            EMConversation conversation = EMClient.getInstance().chatManager().getConversation(conversationId, type, true);
            list.add(conversation);
        }

        EMClient.getInstance().pushManager().getSilentModeForConversations(list, new EMValueWrapperCallBack<Map<String, EMSilentModeResult>>(result, channelName) {
            @Override
            public void onSuccess(Map<String, EMSilentModeResult> object) {
                Map<String ,Map> result = new HashMap<>();
                for (Map.Entry<String, EMSilentModeResult>entry: object.entrySet()) {
                    result.put(entry.getKey(), EMSilentModeResultHelper.toJson(entry.getValue()));
                }
                super.updateObject(result);
            }
        });

    }

    private void setPreferredNotificationLanguage(JSONObject params, String channelName, Result result) throws JSONException {
        String code = params.getString("code");
        EMClient.getInstance().pushManager().setPreferredNotificationLanguage(code, new EMWrapperCallBack(result, channelName, null));
    }

    private void fetchPreferredNotificationLanguage(JSONObject params, String channelName, Result result) throws JSONException {
        EMClient.getInstance().pushManager().getPreferredNotificationLanguage(new EMValueWrapperCallBack<String>(result, channelName){
            @Override
            public void onSuccess(String object) {
                super.onSuccess(object);
            }
        });
    }

    private void setPushTemplate(JSONObject params, String channelName, Result result) throws JSONException {
        String pushTemplateName = params.getString("pushTemplateName");
        EMClient.getInstance().pushManager().setPushTemplate(pushTemplateName, new EMWrapperCallBack(result, channelName, null));
    }

    private void getPushTemplate(JSONObject params, String channelName, Result result) throws JSONException {
        EMClient.getInstance().pushManager().getPushTemplate(new EMValueWrapperCallBack<String>(result, channelName) {
            @Override
            public void onSuccess(String object) {
                super.onSuccess(object);
            }
        });
    }
}

File: ios/Classes/EMPushManagerWrapper.m
//
//  EMPushManagerWrapper.m
//  im_flutter_sdk
//
//  Created by 东海 on 2020/5/7.
//

#import "EMPushManagerWrapper.h"
#import "EMSDKMethod.h"
#import "EMPushOptions+Helper.h"
#import "EMConversation+Helper.h"
#import "EMGroup+Helper.h"
#import "EMSilentModeParam+Helper.h"
#import "EMSilentModeResult+Helper.h"

@implementation EMPushManagerWrapper

- (instancetype)initWithChannelName:(NSString *)aChannelName
                          registrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    if(self = [super initWithChannelName:aChannelName
                               registrar:registrar]) {
    }
    return self;
}

#pragma mark - FlutterPlugin

- (void)handleMethodCall:(FlutterMethodCall*)call
                  result:(FlutterResult)result {
    if ([ChatGetImPushConfig isEqualToString:call.method]) {
        [self getImPushConfig:call.arguments
                  channelName:call.method
                       result:result];
    } else if ([ChatGetImPushConfigFromServer isEqualToString:call.method]) {
        [self getImPushConfigFromServer:call.arguments
                            channelName:call.method
                                 result:result];
    } else if ([ChatUpdatePushNickname isEqualToString:call.method]) {
        [self updatePushNickname:call.arguments
                     channelName:call.method
                          result:result];
    } else if ([ChatUpdateImPushStyle isEqualToString:call.method]) {
        [self updateImPushStyle:call.arguments
                    channelName:call.method
                         result:result];
    } else if ([ChatUpdateGroupPushService isEqualToString:call.method]) {
        [self updateGroupPushService:call.arguments
                         channelName:call.method
                              result:result];
    } else if ([ChatBindDeviceToken isEqualToString:call.method] || [UpdateFCMPushToken isEqualToString:call.method]) {
        [self  bindDeviceToken:call.arguments
                   channelName:call.method
                        result:result];
    } else if ([ChatEnablePush isEqualToString:call.method]) {
        [self enablePush:call.arguments
             channelName:call.method
                  result:result];
    } else if ([ChatDisablePush isEqualToString:call.method]) {
        [self disablePush:call.arguments
              channelName:call.method
                   result:result];
    } else if ([ChatGetNoPushGroups isEqualToString:call.method]) {
        [self getNoPushGroups:call.arguments
                  channelName:call.method
                       result:result];
    } else if ([ChatUpdateUserPushService isEqualToString:call.method]){
        [self updateUserPushService:call.arguments
                        channelName:call.method
                             result:result];
    } else if ([ChatGetNoPushUsers isEqualToString:call.method]){
        [self getNoPushUsers:call.arguments
                 channelName:call.method
                      result:result];
    } else if ([ChatReportPushAction isEqualToString:call.method]){
        [self reportPushAction:call.arguments
                   channelName:call.method
                        result:result];
    } else if ([ChatSetConversationSilentMode isEqualToString:call.method]){
        [self setConversationSilentMode:call.arguments
                            channelName:call.method
                                 result:result];
    } else if ([ChatRemoveConversationSilentMode isEqualToString:call.method]){
        [self removeConversationSilentMode:call.arguments
                               channelName:call.method
                                    result:result];
    } else if ([ChatFetchConversationSilentMode isEqualToString:call.method]){
        [self fetchConversationSilentMode:call.arguments
                              channelName:call.method
                                   result:result];
    } else if ([ChatSetSilentModeForAll isEqualToString:call.method]){
        [self setSilentModeForAll:call.arguments
                      channelName:call.method
                           result:result];
    } else if ([ChatFetchSilentModeForAll isEqualToString:call.method]){
        [self fetchSilentModeForAll:call.arguments
                        channelName:call.method
                             result:result];
    } else if ([ChatFetchSilentModeForConversations isEqualToString:call.method]){
        [self fetchSilentModeForConversations:call.arguments
                                  channelName:call.method
                                       result:result];
    } else if ([ChatSetPreferredNotificationLanguage isEqualToString:call.method]){
        [self setPreferredNotificationLanguage:call.arguments
                                   channelName:call.method
                                        result:result];
    } else if ([ChatFetchPreferredNotificationLanguage isEqualToString:call.method]){
        [self fetchPreferredNotificationLanguage:call.arguments
                                     channelName:call.method
                                          result:result];
    } else if ([ChatSetPushTemplate isEqualToString:call.method]) {
        [self setPushTemplate:call.arguments channelName:call.method result:result];
    } else if ([ChatGetPushTemplate isEqualToString:call.method]) {
        [self getPushTemplate:call.arguments channelName:call.method result:result];
    }
    
    else{
        [super handleMethodCall:call result:result];
    }
}

- (void)getImPushConfig:(NSDictionary *)param
            channelName:(NSString *)aChannelName
                 result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    EMPushOptions *options = EMClient.sharedClient.pushManager.pushOptions;
    [weakSelf wrapperCallBack:result
                  channelName:aChannelName
                        error:nil
                       object:[options toJson]];
}

- (void)getImPushConfigFromServer:(NSDictionary *)param
                      channelName:(NSString *)aChannelName
                           result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    [EMClient.sharedClient.pushManager getPushNotificationOptionsFromServerWithCompletion:^(EMPushOptions *aOptions, EMError *aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:[aOptions toJson]];
        
    }];
}

- (void)updatePushNickname:(NSDictionary *)param
               channelName:(NSString *)aChannelName
                    result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *nickname = param[@"nickname"];
    [EMClient.sharedClient.pushManager updatePushDisplayName:nickname
                                                  completion:^(NSString * _Nonnull aDisplayName, EMError * _Nonnull aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aDisplayName];
    }];
    
}


- (void)updateImPushStyle:(NSDictionary *)param
              channelName:(NSString *)aChannelName
                   result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    
    EMPushDisplayStyle pushStyle = [param[@"pushStyle"] intValue];
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        EMError *aError = [EMClient.sharedClient.pushManager updatePushDisplayStyle:pushStyle];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf wrapperCallBack:result
                          channelName:aChannelName
                                error:aError
                               object:@(!aError)];
        });
    });
}

- (void)updateGroupPushService:(NSDictionary *)param
                   channelName:(NSString *)aChannelName
                        result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSArray *groupIds = param[@"group_ids"];
    bool noPush = [param[@"noPush"] boolValue];
    
    [EMClient.sharedClient.pushManager updatePushServiceForGroups:groupIds
                                                      disablePush:noPush
                                                       completion:^(EMError * _Nonnull aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}


- (void)bindDeviceToken:(NSDictionary *)param
            channelName:(NSString *)aChannelName
                 result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *deviceToken = param[@"token"];
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        EMError *error = [EMClient.sharedClient bindDeviceToken:deviceToken];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf wrapperCallBack:result
                          channelName:aChannelName
                                error:error
                               object:nil];
        });
    });
}


- (void)enablePush:(NSDictionary *)param
       channelName:(NSString *)aChannelName
            result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        EMError *error = [EMClient.sharedClient.pushManager enableOfflinePush];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf wrapperCallBack:result
                          channelName:aChannelName
                                error:error
                               object:nil];
        });
    });
}

- (void)disablePush:(NSDictionary *)param
        channelName:(NSString *)aChannelName
             result:(FlutterResult)result {
    int startTime = [param[@"start"] intValue];
    int endTime = [param[@"end"] intValue];
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        EMError *error = [EMClient.sharedClient.pushManager disableOfflinePushStart:startTime end:endTime];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf wrapperCallBack:result
                          channelName:aChannelName
                                error:error
                               object:nil];
        });
    });
}

- (void)getNoPushGroups:(NSDictionary *)param
            channelName:(NSString *)aChannelName
                 result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSArray<NSString *>* groups = [EMClient.sharedClient.pushManager noPushGroups];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf wrapperCallBack:result
                          channelName:aChannelName
                                error:nil
                               object:groups];
        });
    });
}


- (void)updateUserPushService:(NSDictionary *)param
                  channelName:(NSString *)aChannelName
                       result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSArray *userIds = param[@"user_ids"];
    bool noPush = [param[@"noPush"] boolValue];
    
    [EMClient.sharedClient.pushManager updatePushServiceForUsers:userIds disablePush:noPush completion:^(EMError * _Nonnull aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}

- (void)getNoPushUsers:(NSDictionary *)param
           channelName:(NSString *)aChannelName
                result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        NSArray<NSString *>* userIds = [EMClient.sharedClient.pushManager noPushUIds];
        dispatch_async(dispatch_get_main_queue(), ^{
            [weakSelf wrapperCallBack:result
                          channelName:aChannelName
                                error:nil
                               object:userIds];
        });
    });
}

- (void)reportPushAction:(NSDictionary *)param
             channelName:(NSString *)aChannelName
                  result:(FlutterResult)result {
    
}

- (void)setConversationSilentMode:(NSDictionary *)param
                      channelName:(NSString *)aChannelName
                           result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *conversaionId = param[@"conversationId"];
    EMConversationType type = [EMConversation typeFromInt:[param[@"conversationType"] intValue]];
    EMSilentModeParam *silmentParam = [EMSilentModeParam formJson:param[@"param"]];
    [EMClient.sharedClient.pushManager setSilentModeForConversation:conversaionId
                                                   conversationType:type
                                                             params:silmentParam
                                                         completion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}

- (void)removeConversationSilentMode:(NSDictionary *)param
                         channelName:(NSString *)aChannelName
                              result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *conversaionId = param[@"conversationId"];
    EMConversationType type = [EMConversation typeFromInt:[param[@"conversationType"] intValue]];
    [EMClient.sharedClient.pushManager clearRemindTypeForConversation:conversaionId
                                                     conversationType:type
                                                           completion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}
- (void)fetchConversationSilentMode:(NSDictionary *)param
                        channelName:(NSString *)aChannelName
                             result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *conversaionId = param[@"conversationId"];
    EMConversationType type = [EMConversation typeFromInt:[param[@"conversationType"] intValue]];
    [EMClient.sharedClient.pushManager getSilentModeForConversation:conversaionId
                                                   conversationType:type
                                                         completion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:[aResult toJson]];
    }];
}

- (void)setSilentModeForAll:(NSDictionary *)param
                channelName:(NSString *)aChannelName
                     result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    EMSilentModeParam *silmentParam = [EMSilentModeParam formJson:param[@"param"]];
    [EMClient.sharedClient.pushManager setSilentModeForAll:silmentParam
                                                completion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}

- (void)fetchSilentModeForAll:(NSDictionary *)param
                  channelName:(NSString *)aChannelName
                       result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    [EMClient.sharedClient.pushManager getSilentModeForAllWithCompletion:^(EMSilentModeResult * _Nullable aResult, EMError * _Nullable aError)
     {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:[aResult toJson]];
    }];
}

- (void)fetchSilentModeForConversations:(NSDictionary *)param
                            channelName:(NSString *)aChannelName
                                 result:(FlutterResult)result {
    NSMutableArray *conversations = [NSMutableArray array];
    for (NSString *conversaitonId in param.allKeys) {
        EMConversationType type = [EMConversation typeFromInt:[param[conversaitonId] intValue]];
        EMConversation *conversation = [EMClient.sharedClient.chatManager getConversation:conversaitonId type:type createIfNotExist:YES];
        [conversations addObject:conversation];
    }
    
    __weak typeof(self) weakSelf = self;
    [EMClient.sharedClient.pushManager getSilentModeForConversations:conversations
                                                          completion:^(NSDictionary<NSString *,EMSilentModeResult *> * _Nullable aResult, EMError * _Nullable aError)
     {
        
        NSMutableDictionary *tmpDict = [NSMutableDictionary dictionary];
        for (NSString *conversationId in aResult.allKeys) {
            tmpDict[conversationId] = [aResult[conversationId] toJson];
        }
        
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:tmpDict];
    }];
}

- (void)setPreferredNotificationLanguage:(NSDictionary *)param
                             channelName:(NSString *)aChannelName
                                  result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *code = param[@"code"];
    [EMClient.sharedClient.pushManager setPreferredNotificationLanguage:code completion:^(EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}

- (void)fetchPreferredNotificationLanguage:(NSDictionary *)param
                               channelName:(NSString *)aChannelName
                                    result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    [EMClient.sharedClient.pushManager getPreferredNotificationLanguageCompletion:^(NSString * _Nullable aLaguangeCode, EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aLaguangeCode];
    }];
}

- (void)setPushTemplate:(NSDictionary *)param
                               channelName:(NSString *)aChannelName
                                    result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    NSString *pushTemplateName = param[@"pushTemplateName"];
    [EMClient.sharedClient.pushManager setPushTemplate:pushTemplateName completion:^(EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:nil];
    }];
}

- (void)getPushTemplate:(NSDictionary *)param
                               channelName:(NSString *)aChannelName
                                    result:(FlutterResult)result {
    __weak typeof(self) weakSelf = self;
    [EMClient.sharedClient.pushManager getPushTemplate:^(NSString * _Nullable aPushTemplateName, EMError * _Nullable aError) {
        [weakSelf wrapperCallBack:result
                      channelName:aChannelName
                            error:aError
                           object:aPushTemplateName];
    }];
}

@end
